Assets/Shopify/UIToolkit/Components/VariantSelector.cs

```diff
@@ -0,0 +1,101 @@
+namespace Shopify.UIToolkit {
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Collections.Specialized;
+    using System.Linq;
+    using Shopify.Unity;
+
+    using UnityEngine;
+
+    /// <summary>
+    /// Converts the list of product variant and their respective selectable options that looks like:
+    /// [{
+    ///    "selectedOptions": [{
+    ///       "name": "optionName1",
+    ///       "value": "value1"
+    ///     }, {
+    ///       "name": "optionName2",
+    ///       "value": "value1"
+    ///     }]
+    /// },
+    /// {
+    ///    "selectedOptions": [{
+    ///       "name": "optionName1",
+    ///       "value": "value2"
+    ///     }, {
+    ///       "name": "optionName2",
+    ///       "value": "value2"
+    ///     }]
+    /// }]
+    /// into a map that looks like:
+    /// {
+    ///   "optionName1": ["value1", "value2", ...],
+    ///   "optionName2": ["value1", "value2", ...],
+    /// }
+    ///
+    /// </summary>
+    public class VariantSelector {
+        private List<ProductVariant> _variants;
+
+        public VariantSelector(List<ProductVariant> variants) {
+            _variants = variants;
+        }
+
+        public Dictionary<string, List<string>> AllOptions() {
+            return ValidOptionsForSelections(new Dictionary<string, string>());
+        }
+
+        public Dictionary<string, List<string>> ValidOptionsForSelections(Dictionary<string, string> selections) {
+                if (IsSingleOrEmpty()) {
+                    return null;
+                }
+
+                var allOptions = new Dictionary<string, List<string>>();
+                foreach (var variant in _variants) {
+                    var options = variant.selectedOptions();
+
+                    var containsSelections = selections.Aggregate(true, (accum, entry) => {
+                        return accum && options.Find(o => {
+                            return o.name().Equals(entry.Key) && o.value().Equals(entry.Value);
+                        }) != null;
+                    });
+
+                    // Skip this variant if it's not valid with the selections.
+                    if (!containsSelections) continue;
+
+                    options.Aggregate(allOptions, (accum, option) => {
+                        List<string> values;
+                        if (accum.ContainsKey(option.name())) {
```

Comment:
This can just be `if (!accum.ContainsKey(option.name())) {` then outside of the if statement do `values = accum[option.name()];`